cmd/bench.go

```diff
@@ -0,0 +1,300 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"sort"
+	"testing"
+
+	"github.com/olekukonko/tablewriter"
+	"github.com/spf13/cobra"
+
+	"github.com/open-policy-agent/opa/internal/presentation"
+
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/util"
+)
+
+// benchmarkCommandParams are a superset of evalCommandParams
+// but not all eval options are exposed with flags. Only the
+// ones compatible with running a benchmark.
+type benchmarkCommandParams struct {
+	evalCommandParams
+	benchMem bool
+	count    int
+}
+
+const (
+	benchmarkGoBenchOutput = "gobench"
+)
+
+func newBenchmarkEvalParams() benchmarkCommandParams {
+	return benchmarkCommandParams{
+		evalCommandParams: evalCommandParams{
+			outputFormat: util.NewEnumFlag(evalPrettyOutput, []string{
+				evalJSONOutput,
+				evalPrettyOutput,
+				benchmarkGoBenchOutput,
+			}),
+		},
+	}
+}
+
+func init() {
+	params := newBenchmarkEvalParams()
+
+	benchCommand := &cobra.Command{
+		Use:   "bench <query>",
+		Short: "Benchmark a Rego query",
+		Long: `Benchmark a Rego query and print the results.
+
+The benchmark command works very similar to 'eval' and will evaluate the query in the same fashion. The
+evaluation will be repeated a number of times and performance results will be returned.
+		
+Example with bundle and input data:
+		
+	opa bench -b ./policy-bundle -i input.json 'data.authz.allow'
+
+
+To enable more detailed analysis use the --metrics and --benchmem flags.
+		
+The optional "gobench" output format conforms to the Go Benchmark Data Format.
+`,
+
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return validateEvalParams(&params.evalCommandParams, args)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			os.Exit(benchMain(args, params, os.Stdout, &goBenchRunner{}))
+		},
+	}
+
+	// Sub-set of the standard `opa eval ..` flags
+	addFailFlag(benchCommand.Flags(), &params.fail, true)
+	addDataFlag(benchCommand.Flags(), &params.dataPaths)
+	addBundleFlag(benchCommand.Flags(), &params.bundlePaths)
+	addInputFlag(benchCommand.Flags(), &params.inputPath)
+	addImportFlag(benchCommand.Flags(), &params.imports)
+	addPackageFlag(benchCommand.Flags(), &params.pkg)
+	addQueryStdinFlag(benchCommand.Flags(), &params.stdin)
+	addInputStdinFlag(benchCommand.Flags(), &params.stdinInput)
+	addMetricsFlag(benchCommand.Flags(), &params.metrics, true)
+	addOutputFormat(benchCommand.Flags(), params.outputFormat)
+	addIgnoreFlag(benchCommand.Flags(), &params.ignore)
+
+	// Shared benchmark flags
+	addCountFlag(benchCommand.Flags(), &params.count, "benchmark")
+	addBenchmemFlag(benchCommand.Flags(), &params.benchMem, true)
+
+	RootCommand.AddCommand(benchCommand)
+}
+
+type benchRunner interface {
```

Comment:
NOTE: not sure why we have this interface yet.

Reply:
Makes testing this wayyyy easier. We don't have much control over the golang benchmark suite to like crank down `b.N` to small numbers so having a way to mock out the actual runner helps make the unit tests much faster.

Reply:
👍 thanks for clarifying.